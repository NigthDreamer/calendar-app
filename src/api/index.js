//* Renombramos la exportacion por defecto
export { default as calendarApi } from './calendarApi';
