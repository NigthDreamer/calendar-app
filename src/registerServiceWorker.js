if ('serviceWorker' in navigator) {
  window.addEventListener('load', () => {
    navigator.serviceWorker.register('/service.js', {scope: '/'})
      .then(registration => {
        console.log('Service Worker registered with scope:', registration.scope);
      })
      .catch(error => {
        console.error('Service Worker registration failed:', error);
      });
  });
}