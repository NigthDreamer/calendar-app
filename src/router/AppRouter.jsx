import { RouterProvider, createBrowserRouter, Navigate } from 'react-router-dom';
import { AuthRoutes } from '../auth';
import { CalendarRoutes } from '../calendar';

const router = createBrowserRouter([
  {
    path: '*',
    element: <Navigate to="auth" replace />,
  },
  AuthRoutes,
  CalendarRoutes,
]);

export const AppRouter = () => {
  return <RouterProvider router={router} />;
};
