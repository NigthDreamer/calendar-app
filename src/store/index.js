export * from './auth';
export * from './calendar';
export * from './store';
export * from './ui';