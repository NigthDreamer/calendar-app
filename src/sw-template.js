/* eslint-disable no-undef */
importScripts('https://storage.googleapis.com/workbox-cdn/releases/6.4.1/workbox-sw.js');

workbox.loadModule('workbox-background-sync');

workbox.precaching.precacheAndRoute(self.__WB_MANIFEST);

const { registerRoute } = workbox.routing;
const { CacheFirst, NetworkFirst, NetworkOnly } = workbox.strategies;
const { BackgroundSyncPlugin } = workbox.backgroundSync;

const networkFirst = [
  '/api/auth/renew',
  '/api/events',
]

const cacheFirst = [
  'https://kit.fontawesome.com/caf6c2f77a.js'
]

const networkOnly = [
  '/api/events',
  '/api/events/',
]

registerRoute(
  ({url}) => {
    return networkFirst.includes(url.pathname);
  },
	new NetworkFirst()
);

registerRoute(
  ({url}) => {
    return cacheFirst.includes(url.href);
  },
	//* Aplicar la estrategia que se desee
	new CacheFirst()
);

// Offline Posts
const bgSyncPlugin = new BackgroundSyncPlugin('posteos-offline', {
  maxRetentionTime: 24 * 60, // Retry for max of 24 Hours (specified in minutes)
});

const statusPlugin = {
  fetchDidSucceed: ({response}) => {
    if (response.status >= 500) {
      // Throwing anything here will trigger fetchDidFail.
      throw new Error('Server error.');
    }
    // If it's not 5xx, use the response as-is.
    return response;
  },
};

registerRoute(
  ({url}) => {
    return networkOnly.includes(url.pathname);
  },
	new NetworkOnly({
    plugins: [bgSyncPlugin, statusPlugin]
  }),
	'POST'
);

registerRoute(
  ({url}) => {
    return networkOnly.includes(url.pathname);
  },
	new NetworkOnly({
    plugins: [bgSyncPlugin, statusPlugin]
  }),
	'DELETE'
);

registerRoute(
  ({url}) => {
    return networkOnly.includes(url.pathname);
  },
	new NetworkOnly({
    plugins: [bgSyncPlugin, statusPlugin]
  }),
	'PUT'
);