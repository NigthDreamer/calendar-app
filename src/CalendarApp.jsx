import { Provider } from 'react-redux';
import { AppRouter } from './router/';
import { store } from './store';

export const CalendarApp = () => {
  return (
    <Provider store={store}>
      <AppRouter />
    </Provider>
  );
};
