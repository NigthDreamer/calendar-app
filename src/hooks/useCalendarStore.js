import { useSelector, useDispatch } from 'react-redux';
import { onAddNewEvent, onDeleteEvent, onSetActiveEvent, onUpdateEvent } from '../store';

export const useCalendarStore = () => {
  const dispatch = useDispatch();
  const { events, activeEvent } = useSelector((state) => state.calendar);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent) => {
    if (calendarEvent._id) {
      // Actualizando
      dispatch(onUpdateEvent({...calendarEvent}));
    } else {
      // Creando
      dispatch(onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }));
    }
  };

  const startDeletingEvent = () => {
    // Todo: Llegar al backend
    dispatch(onDeleteEvent());
  } 

  return {
    //* Propiedades
    events,
    activeEvent,
    hasEventSelected: !!activeEvent,
    isNewNote: !!activeEvent && !activeEvent._id,
    //* MÃ©todos
    setActiveEvent,
    startSavingEvent,
    startDeletingEvent,
  };
};
