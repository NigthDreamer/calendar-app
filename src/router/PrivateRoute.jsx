import { Navigate } from 'react-router-dom';
import { useAuthStore } from '../hooks';
import { useEffect } from 'react';

//Al poner el children, definimos este componente como un higher order component
// eslint-disable-next-line react/prop-types
export const PrivateRoute = ({ children }) => {

  const { status, checkAuthToken } = useAuthStore();

  useEffect(() => {
    checkAuthToken();
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if (status === 'checking') {
    return <h3>Cargando...</h3>;
  }

  return status === 'authenticated' ? (
    children
  ) : (
    <Navigate to="/auth/login" replace={true} />
  );
};
