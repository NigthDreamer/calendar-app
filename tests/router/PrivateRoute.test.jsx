import { render, screen } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { useAuthStore } from '../../src/hooks/useAuthStore';
import { PrivateRoute } from '../../src/router/PrivateRoute';
import { LoginPage } from '../../src/auth';
import { CalendarPage } from '../../src/calendar';

/* eslint-disable no-undef */
jest.mock('../../src/hooks/useAuthStore');

/**
 * Hago un mock del componente CalendarPage para evitar tener que renderizar
 * el componente completo, lo que implica mockear el store, todos los hooks que
 * use, inicializar el estado de los slices... etc.
 */
jest.mock('../../src/calendar', () => ({
  CalendarPage: () => <h1>CalendarPage</h1>
}));

describe('Pruebas en <AppRouter/>', () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Debe de mostrar la pantalla de carga y llamar checkAuthToken', () => {
    useAuthStore.mockReturnValue({
      status: 'checking',
      checkAuthToken: mockCheckAuthToken,
    });

    render(<PrivateRoute/>);

    expect(screen.getByText('Cargando...')).toBeTruthy();
    expect(mockCheckAuthToken).toHaveBeenCalled();
  });

  test('Debe de mostrar el login en caso de no estar autenticado', () => {
    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkAuthToken: mockCheckAuthToken,
    });

    /**
     * Hay que definir las rutas que vamos a probar en el MemoryRouter. En este caso
     * la ruta calendar representara una ruta privada y auth/login la ruta publica.
     * Al navegar a calendar, al ser una ruta privada y como nuestro status es 
     * 'not-authenticated', nos redireccionar√° a auth/login
     */
    const {container} = render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={
              <PrivateRoute>
                <h1>Ruta Privada</h1>
              </PrivateRoute>
            }
          />
          <Route path="auth/login" element={<LoginPage/>} />
        </Routes>
      </MemoryRouter>
    );

    screen.debug();
    expect(screen.getByText('Ingreso')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
  
  test('Debe de mostrar el calendario si estamos autenticados', () => {
    useAuthStore.mockReturnValue({
      status: 'authenticated',
      checkAuthToken: mockCheckAuthToken,
    });

    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={
              <PrivateRoute>
                {/* Este componente esta mockeado */}
                <CalendarPage/> 
              </PrivateRoute>
            }
          />
          <Route path="auth/login" element={<LoginPage/>} />
        </Routes>
      </MemoryRouter>
    );

    expect(screen.getByText('CalendarPage')).toBeTruthy();
  });
});
