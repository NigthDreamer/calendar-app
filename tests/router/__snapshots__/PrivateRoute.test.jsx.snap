// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter/> Debe de mostrar el login en caso de no estar autenticado 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              autocomplete="email"
              class="form-control"
              name="loginEmail"
              placeholder="Correo"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              autocomplete="current-password"
              class="form-control"
              name="loginPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1agp3zg-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              id=":r0:"
              tabindex="0"
              type="submit"
            >
              Login
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              autocomplete="username"
              class="form-control"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              autocomplete="email"
              class="form-control"
              name="registerEmail"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              autocomplete="new-password"
              class="form-control"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              autocomplete="new-password"
              class="form-control"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInherit MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedInherit MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInherit css-mkie7h-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              id=":r1:"
              tabindex="0"
              type="submit"
            >
              Crear cuenta
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
